File: MetalSampleCodeLibrary/LightingTechniques/RenderingASceneWithDeferredLightingInCplusplus/Renderer/AAPLRenderer_TraditionalDeferred.cpp
/*
See LICENSE folder for this sample’s licensing information.

Abstract:
Implementation of the renderer class that performs Metal setup and per-frame rendering for a
 traditional deferred renderer used for macOS devices without Apple silicon and the
 iOS & tvOS simulators.
*/

#include "AAPLRenderer_TraditionalDeferred.h"

// Include header shared between C code here, which executes Metal API commands, and .metal files.
#include "AAPLShaderTypes.h"

#include "AAPLUtilities.h"

Renderer_TraditionalDeferred::Renderer_TraditionalDeferred(MTL::Device* pDevice)
: Renderer( pDevice )
{
    m_singlePassDeferred = false;
    loadMetalInternal();
    loadScene();
}

Renderer_TraditionalDeferred::~Renderer_TraditionalDeferred()
{
    m_pLightPipelineState->release();
    m_pGBufferRenderPassDescriptor->release();
    m_pFinalRenderPassDescriptor->release();
}

/// Create traditional deferred renderer specific Metal state objects
void Renderer_TraditionalDeferred::loadMetalInternal()
{
    Renderer::loadMetal();

    NS::Error* pError = nullptr;


    #pragma mark Point light render pipeline setup
    {
        MTL::RenderPipelineDescriptor* pRenderPipelineDescriptor = MTL::RenderPipelineDescriptor::alloc()->init();

        pRenderPipelineDescriptor->setLabel( AAPLSTR( "Light" ) );
        pRenderPipelineDescriptor->colorAttachments()->object(RenderTargetLighting)->setPixelFormat( colorTargetPixelFormat() );

        // Enable additive blending
        pRenderPipelineDescriptor->colorAttachments()->object(RenderTargetLighting)->setBlendingEnabled( true );
        pRenderPipelineDescriptor->colorAttachments()->object(RenderTargetLighting)->setRgbBlendOperation( MTL::BlendOperationAdd );
        pRenderPipelineDescriptor->colorAttachments()->object(RenderTargetLighting)->setAlphaBlendOperation( MTL::BlendOperationAdd );
        pRenderPipelineDescriptor->colorAttachments()->object(RenderTargetLighting)->setDestinationRGBBlendFactor( MTL::BlendFactorOne );
        pRenderPipelineDescriptor->colorAttachments()->object(RenderTargetLighting)->setDestinationAlphaBlendFactor( MTL::BlendFactorOne );
        pRenderPipelineDescriptor->colorAttachments()->object(RenderTargetLighting)->setSourceRGBBlendFactor( MTL::BlendFactorOne );
        pRenderPipelineDescriptor->colorAttachments()->object(RenderTargetLighting)->setSourceAlphaBlendFactor( MTL::BlendFactorOne );

        pRenderPipelineDescriptor->setDepthAttachmentPixelFormat( depthStencilTargetPixelFormat() );
        pRenderPipelineDescriptor->setStencilAttachmentPixelFormat( depthStencilTargetPixelFormat() );

        MTL::Library* pShaderLibrary = m_pDevice->newDefaultLibrary();

        AAPL_ASSERT( pShaderLibrary, "Failed to create default shader library" );

        MTL::Function* pLightVertexFunction = pShaderLibrary->newFunction( AAPLSTR( "deferred_point_lighting_vertex" ) );
        MTL::Function* pLightFragmentFunction = pShaderLibrary->newFunction( AAPLSTR( "deferred_point_lighting_fragment_traditional" ) );

        AAPL_ASSERT( pLightVertexFunction, "Failed to load deferred_point_lighting_vertex" );
        AAPL_ASSERT( pLightFragmentFunction, "Failed to load deferred_point_lighting_fragment_traditional" );

        pRenderPipelineDescriptor->setVertexFunction( pLightVertexFunction );
        pRenderPipelineDescriptor->setFragmentFunction( pLightFragmentFunction );

        m_pLightPipelineState = m_pDevice->newRenderPipelineState( pRenderPipelineDescriptor, &pError );

        AAPL_ASSERT_NULL_ERROR( pError, "Failed to create lighting render pipeline state" );
        
        pLightVertexFunction->release();
        pLightFragmentFunction->release();
        pRenderPipelineDescriptor->release();
        pShaderLibrary->release();
    }

    #pragma mark GBuffer render pass descriptor setup
    // Create a render pass descriptor to create an encoder for rendering to the GBuffers.
    // The encoder stores rendered data of each attachment when encoding ends.
    m_pGBufferRenderPassDescriptor = MTL::RenderPassDescriptor::alloc()->init();
    
    m_pGBufferRenderPassDescriptor->colorAttachments()->object(RenderTargetLighting)->setLoadAction( MTL::LoadActionDontCare );
    m_pGBufferRenderPassDescriptor->colorAttachments()->object(RenderTargetLighting)->setStoreAction( MTL::StoreActionDontCare );
    m_pGBufferRenderPassDescriptor->colorAttachments()->object(RenderTargetAlbedo)->setLoadAction( MTL::LoadActionDontCare );
    m_pGBufferRenderPassDescriptor->colorAttachments()->object(RenderTargetAlbedo)->setStoreAction( MTL::StoreActionStore );
    m_pGBufferRenderPassDescriptor->colorAttachments()->object(RenderTargetNormal)->setLoadAction( MTL::LoadActionDontCare );
    m_pGBufferRenderPassDescriptor->colorAttachments()->object(RenderTargetNormal)->setStoreAction( MTL::StoreActionStore );
    m_pGBufferRenderPassDescriptor->colorAttachments()->object(RenderTargetDepth)->setLoadAction( MTL::LoadActionDontCare );
    m_pGBufferRenderPassDescriptor->colorAttachments()->object(RenderTargetDepth)->setStoreAction( MTL::StoreActionStore );
    m_pGBufferRenderPassDescriptor->depthAttachment()->setClearDepth( 1.0 );
    m_pGBufferRenderPassDescriptor->depthAttachment()->setLoadAction( MTL::LoadActionClear );
    m_pGBufferRenderPassDescriptor->depthAttachment()->setStoreAction( MTL::StoreActionStore );

    m_pGBufferRenderPassDescriptor->stencilAttachment()->setClearStencil( 0 );
    m_pGBufferRenderPassDescriptor->stencilAttachment()->setLoadAction( MTL::LoadActionClear );
    m_pGBufferRenderPassDescriptor->stencilAttachment()->setStoreAction( MTL::StoreActionStore );

    // Create a render pass descriptor for the lighting and composition pass

    // Whatever rendered in the final pass needs to be stored so it can be displayed
    m_pFinalRenderPassDescriptor = MTL::RenderPassDescriptor::alloc()->init();
    m_pFinalRenderPassDescriptor->colorAttachments()->object(0)->setStoreAction( MTL::StoreActionStore );
    m_pFinalRenderPassDescriptor->depthAttachment()->setLoadAction( MTL::LoadActionLoad );
    m_pFinalRenderPassDescriptor->stencilAttachment()->setLoadAction( MTL::LoadActionLoad );
}

/// Respond to view size change
void Renderer_TraditionalDeferred::drawableSizeWillChange(const MTL::Size& size, MTL::StorageMode GBufferStorageMode)
{
    // The renderer base class allocates all GBuffers >except< lighting GBuffer (since with the
    // single-pass deferred renderer the lighting buffer is the same as the drawable)
    Renderer::drawableSizeWillChange( size, GBufferStorageMode );

    // Re-set GBuffer textures in the GBuffer render pass descriptor after they have been
    // reallocated by a resize
    m_pGBufferRenderPassDescriptor->colorAttachments()->object(RenderTargetAlbedo)->setTexture( m_albedo_specular_GBuffer );
    m_pGBufferRenderPassDescriptor->colorAttachments()->object(RenderTargetNormal)->setTexture( m_normal_shadow_GBuffer );
    m_pGBufferRenderPassDescriptor->colorAttachments()->object(RenderTargetDepth)->setTexture( m_depth_GBuffer );
}

/// Draw directional lighting, which, with a tradition deferred renderer needs to set GBuffers as
/// textures before executing common rendering code to draw the light
void Renderer_TraditionalDeferred::drawDirectionalLight(MTL::RenderCommandEncoder* pRenderEncoder)
{
    pRenderEncoder->pushDebugGroup( AAPLSTR( "Draw Directional Light" ) );
    pRenderEncoder->setFragmentTexture( m_albedo_specular_GBuffer, RenderTargetAlbedo );
    pRenderEncoder->setFragmentTexture( m_normal_shadow_GBuffer, RenderTargetNormal );
    pRenderEncoder->setFragmentTexture( m_depth_GBuffer, RenderTargetDepth );

    Renderer::drawDirectionalLightCommon( pRenderEncoder );

    pRenderEncoder->popDebugGroup();
}

/// Setup traditional deferred rendering specific pipeline and set GBuffer textures.  Then call
/// common renderer code to apply the point lights
void Renderer_TraditionalDeferred::drawPointLights(MTL::RenderCommandEncoder* pRenderEncoder)
{
    pRenderEncoder->pushDebugGroup( AAPLSTR( "Draw Point Lights" ) );

    pRenderEncoder->setRenderPipelineState( m_pLightPipelineState );

    pRenderEncoder->setFragmentTexture( m_albedo_specular_GBuffer, RenderTargetAlbedo );
    pRenderEncoder->setFragmentTexture( m_normal_shadow_GBuffer, RenderTargetNormal );
    pRenderEncoder->setFragmentTexture( m_depth_GBuffer, RenderTargetDepth );

    // Call common base class method after setting state in the renderEncoder specific to the
    // traditional deferred renderer
    Renderer::drawPointLightsCommon( pRenderEncoder );

    pRenderEncoder->popDebugGroup();
}

/// Frame drawing routine
void Renderer_TraditionalDeferred::drawInView( bool isPaused, MTL::Drawable* pCurrentDrawable, MTL::Texture* pDepthStencilTexture )
{
    {
        MTL::CommandBuffer* pCommandBuffer = Renderer::beginFrame( isPaused );
        pCommandBuffer->setLabel( AAPLSTR( "Shadow & GBuffer Commands" ) );

        Renderer::drawShadow( pCommandBuffer );

        m_pGBufferRenderPassDescriptor->depthAttachment()->setTexture( pDepthStencilTexture );
        m_pGBufferRenderPassDescriptor->stencilAttachment()->setTexture( pDepthStencilTexture );

        MTL::RenderCommandEncoder* pRenderEncoder = pCommandBuffer->renderCommandEncoder( m_pGBufferRenderPassDescriptor );
        pRenderEncoder->setLabel( AAPLSTR( "GBuffer Generation" ) );

        Renderer::drawGBuffer( pRenderEncoder );

        pRenderEncoder->endEncoding();

        // Commit commands so that Metal can begin working on nondrawable dependent work without
        // waiting for a drawable to become available
        pCommandBuffer->commit();
    }

    {
        MTL::CommandBuffer* pCommandBuffer = Renderer::beginDrawableCommands();

        pCommandBuffer->setLabel( AAPLSTR( "Lighting Commands" ) );

        MTL::Texture* pDrawableTexture = Renderer::currentDrawableTexture( pCurrentDrawable );

        // The final pass can only render if a drawable is available, otherwise it needs to skip
        // rendering this frame.
        if( pDrawableTexture )
        {
            // Render the lighting and composition pass

            m_pFinalRenderPassDescriptor->colorAttachments()->object(0)->setTexture( pDrawableTexture );
            m_pFinalRenderPassDescriptor->depthAttachment()->setTexture( pDepthStencilTexture );
            m_pFinalRenderPassDescriptor->stencilAttachment()->setTexture( pDepthStencilTexture );

            MTL::RenderCommandEncoder* pRenderEncoder = pCommandBuffer->renderCommandEncoder( m_pFinalRenderPassDescriptor );
            pRenderEncoder->setLabel( AAPLSTR( "Lighting & Composition Pass" ) );

            drawDirectionalLight( pRenderEncoder );

            Renderer::drawPointLightMask( pRenderEncoder );

            drawPointLights( pRenderEncoder );

            Renderer::drawSky( pRenderEncoder );

            Renderer::drawFairies( pRenderEncoder );

            pRenderEncoder->endEncoding();
        }

        Renderer::endFrame( pCommandBuffer, pCurrentDrawable );
    }
}

#if SUPPORT_BUFFER_EXAMINATION

/// Set up render targets for display when buffer examination mode enabled. Set up target for
/// optimal rendering when buffer examination mode disabled.
void Renderer_TraditionalDeferred::validateBufferExaminationMode()
{
    if( m_bufferExaminationManager->mode() )
    {
        // Clear the background of the GBuffer when examining buffers.  When rendering normally
        // clearing is wasteful, but when examining the buffers, the backgrounds appear corrupt
        // making unclear what's actually rendered to the buffers
        m_pGBufferRenderPassDescriptor->colorAttachments()->object(RenderTargetAlbedo)->setLoadAction( MTL::LoadActionClear );
        m_pGBufferRenderPassDescriptor->colorAttachments()->object(RenderTargetNormal)->setLoadAction( MTL::LoadActionClear );
        m_pGBufferRenderPassDescriptor->colorAttachments()->object(RenderTargetDepth)->setLoadAction( MTL::LoadActionClear );

        // Store depth and stencil buffers after filling them.  This is wasteful when rendering
        // normally, but necessary to present the light mask culling view.
        m_pFinalRenderPassDescriptor->stencilAttachment()->setStoreAction( MTL::StoreActionStore );
        m_pFinalRenderPassDescriptor->depthAttachment()->setStoreAction( MTL::StoreActionStore );
    }
    else
    {
        // When exiting buffer examination mode, return to efficient state settings
        m_pFinalRenderPassDescriptor->stencilAttachment()->setStoreAction( MTL::StoreActionDontCare );
        m_pFinalRenderPassDescriptor->depthAttachment()->setStoreAction( MTL::StoreActionDontCare );
        m_pGBufferRenderPassDescriptor->colorAttachments()->object(RenderTargetAlbedo)->setLoadAction( MTL::LoadActionDontCare );
        m_pGBufferRenderPassDescriptor->colorAttachments()->object(RenderTargetNormal)->setLoadAction( MTL::LoadActionDontCare );
        m_pGBufferRenderPassDescriptor->colorAttachments()->object(RenderTargetDepth)->setLoadAction( MTL::LoadActionDontCare );
    }
}

#endif // END SUPPORT_BUFFER_EXAMINATION 

File: MetalSampleCodeLibrary/LightingTechniques/RenderingASceneWithDeferredLightingInObjectiveC/Renderer/AAPLRenderer_TraditionalDeferred.m
/*
See LICENSE folder for this sample’s licensing information.

Abstract:
Implementation of the renderer class which performs Metal setup and per frame rendering for a
traditional deferred renderer used for macOS and the iOS & tvOS simulators
*/

#import "AAPLRenderer_TraditionalDeferred.h"

// Include header shared between C code here, which executes Metal API commands, and .metal files
#import "AAPLShaderTypes.h"

@implementation AAPLRenderer_TraditionalDeferred
{
    id <MTLRenderPipelineState> _lightPipelineState;

    MTLRenderPassDescriptor *_GBufferRenderPassDescriptor;
    MTLRenderPassDescriptor *_finalRenderPassDescriptor;
}

/// Perform traditional deferred renderer specific initialization
- (nonnull instancetype)initWithMetalKitView:(nonnull MTKView *)view
{
    self = [super initWithMetalKitView:view];

    if(self)
    {
        self.singlePassDeferred = NO;
        [self loadMetal];
        [self loadScene];
    }

    return self;
}

/// Create traditional deferred renderer specific Metal state objects
- (void)loadMetal
{
    [super loadMetal];

    NSError *error;

    id <MTLLibrary> shaderLibrary = self.shaderLibrary;

    #pragma mark Point light render pipeline setup
    {
        MTLRenderPipelineDescriptor * renderPipelineDescriptor = [MTLRenderPipelineDescriptor new];
        renderPipelineDescriptor.colorAttachments[AAPLRenderTargetLighting].pixelFormat = self.view.colorPixelFormat;

        // Enable additive blending
        renderPipelineDescriptor.colorAttachments[AAPLRenderTargetLighting].blendingEnabled = YES;
        renderPipelineDescriptor.colorAttachments[AAPLRenderTargetLighting].rgbBlendOperation = MTLBlendOperationAdd;
        renderPipelineDescriptor.colorAttachments[AAPLRenderTargetLighting].alphaBlendOperation = MTLBlendOperationAdd;
        renderPipelineDescriptor.colorAttachments[AAPLRenderTargetLighting].destinationRGBBlendFactor = MTLBlendFactorOne;
        renderPipelineDescriptor.colorAttachments[AAPLRenderTargetLighting].destinationAlphaBlendFactor = MTLBlendFactorOne;
        renderPipelineDescriptor.colorAttachments[AAPLRenderTargetLighting].sourceRGBBlendFactor = MTLBlendFactorOne;
        renderPipelineDescriptor.colorAttachments[AAPLRenderTargetLighting].sourceAlphaBlendFactor = MTLBlendFactorOne;

        renderPipelineDescriptor.depthAttachmentPixelFormat = self.view.depthStencilPixelFormat;
        renderPipelineDescriptor.stencilAttachmentPixelFormat = self.view.depthStencilPixelFormat;

        id <MTLFunction> lightVertexFunction = [shaderLibrary newFunctionWithName:@"deferred_point_lighting_vertex"];
        id <MTLFunction> lightFragmentFunction = [shaderLibrary newFunctionWithName:@"deferred_point_lighting_fragment_traditional"];

        renderPipelineDescriptor.label = @"Light";
        renderPipelineDescriptor.vertexFunction = lightVertexFunction;
        renderPipelineDescriptor.fragmentFunction = lightFragmentFunction;
        _lightPipelineState = [self.device newRenderPipelineStateWithDescriptor:renderPipelineDescriptor
                                                                      error:&error];

        NSAssert(_lightPipelineState, @"Failed to create lighting render pipeline state: %@", error);
    }

    #pragma mark GBuffer render pass descriptor setup
    // Create a render pass descriptor to create an encoder for rendering to the GBuffers.
    // The encoder stores rendered data of each attachment when encoding ends.
    _GBufferRenderPassDescriptor = [MTLRenderPassDescriptor new];

    _GBufferRenderPassDescriptor.colorAttachments[AAPLRenderTargetLighting].loadAction = MTLLoadActionDontCare;
    _GBufferRenderPassDescriptor.colorAttachments[AAPLRenderTargetLighting].storeAction = MTLStoreActionDontCare;
    _GBufferRenderPassDescriptor.colorAttachments[AAPLRenderTargetAlbedo].loadAction = MTLLoadActionDontCare;
    _GBufferRenderPassDescriptor.colorAttachments[AAPLRenderTargetAlbedo].storeAction = MTLStoreActionStore;
    _GBufferRenderPassDescriptor.colorAttachments[AAPLRenderTargetNormal].loadAction = MTLLoadActionDontCare;
    _GBufferRenderPassDescriptor.colorAttachments[AAPLRenderTargetNormal].storeAction = MTLStoreActionStore;
    _GBufferRenderPassDescriptor.colorAttachments[AAPLRenderTargetDepth].loadAction = MTLLoadActionDontCare;
    _GBufferRenderPassDescriptor.colorAttachments[AAPLRenderTargetDepth].storeAction = MTLStoreActionStore;
    _GBufferRenderPassDescriptor.depthAttachment.clearDepth = 1.0;
    _GBufferRenderPassDescriptor.depthAttachment.loadAction = MTLLoadActionClear;
    _GBufferRenderPassDescriptor.depthAttachment.storeAction = MTLStoreActionStore;

    _GBufferRenderPassDescriptor.stencilAttachment.clearStencil = 0;
    _GBufferRenderPassDescriptor.stencilAttachment.loadAction = MTLLoadActionClear;
    _GBufferRenderPassDescriptor.stencilAttachment.storeAction = MTLStoreActionStore;

    // Create a render pass descriptor for thelighting and composition pass
    _finalRenderPassDescriptor = [MTLRenderPassDescriptor new];

    // Whatever rendered in the final pass needs to be stored so it can be displayed
    _finalRenderPassDescriptor.colorAttachments[0].storeAction = MTLStoreActionStore;
    _finalRenderPassDescriptor.depthAttachment.loadAction = MTLLoadActionLoad;
    _finalRenderPassDescriptor.stencilAttachment.loadAction = MTLLoadActionLoad;
}

/// MTKViewDelegate Callback: Respond to device orientation change or other view size change
- (void) mtkView:(nonnull MTKView *)view drawableSizeWillChange:(CGSize)size
{
    // The renderer base class allocates all GBuffers >except< lighting GBuffer (since with the
    // single-pass deferred renderer the lighting buffer is the same as the drawable)
    [self drawableSizeWillChange:size withGBufferStorageMode:MTLStorageModePrivate];

    // Re-set GBuffer textures in the GBuffer render pass descriptor after they have been
    // reallocated by a resize
    _GBufferRenderPassDescriptor.colorAttachments[AAPLRenderTargetAlbedo].texture = self.albedo_specular_GBuffer;
    _GBufferRenderPassDescriptor.colorAttachments[AAPLRenderTargetNormal].texture = self.normal_shadow_GBuffer;
    _GBufferRenderPassDescriptor.colorAttachments[AAPLRenderTargetDepth].texture = self.depth_GBuffer;

    // Cannot set the depth stencil texture here since MTKView reallocates it *after* the
    // drawableSizeWillChange callback

    if(view.paused)
    {
        [view draw];
    }
}



/// Draw directional lighting, which, with a tradition deferred renderer needs to set GBuffers as
/// textures before executing common rendering code to draw the light
- (void)drawDirectionalLight:(nonnull id <MTLRenderCommandEncoder>)renderEncoder
{
    [renderEncoder pushDebugGroup:@"Draw Directional Light"];
    [renderEncoder setFragmentTexture:self.albedo_specular_GBuffer atIndex:AAPLRenderTargetAlbedo];
    [renderEncoder setFragmentTexture:self.normal_shadow_GBuffer atIndex:AAPLRenderTargetNormal];
    [renderEncoder setFragmentTexture:self.depth_GBuffer atIndex:AAPLRenderTargetDepth];

    [super drawDirectionalLightCommon:renderEncoder];

    [renderEncoder popDebugGroup];
}

/// Setup tradition deferred rendering specific pipeline and set GBuffer textures. Then call common
/// renderer code to apply the point lights
- (void) drawPointLights:(id<MTLRenderCommandEncoder>)renderEncoder
{
    [renderEncoder pushDebugGroup:@"Draw Point Lights"];

    [renderEncoder setRenderPipelineState:_lightPipelineState];

    [renderEncoder setFragmentTexture:self.albedo_specular_GBuffer atIndex:AAPLRenderTargetAlbedo];
    [renderEncoder setFragmentTexture:self.normal_shadow_GBuffer atIndex:AAPLRenderTargetNormal];
    [renderEncoder setFragmentTexture:self.depth_GBuffer atIndex:AAPLRenderTargetDepth];

    // Call common base class method after setting state in the renderEncoder specific to the
    // traditional deferred renderer
    [super drawPointLightsCommon:renderEncoder];

    [renderEncoder popDebugGroup];
}

/// MTKViewDelegate callback: Called whenever the view needs to render
- (void)drawSceneToView:(nonnull MTKView *)view
{
    id<MTLCommandBuffer> commandBuffer = [self beginFrame];
    commandBuffer.label = @"Shadow & GBuffer Commands";

    [super drawShadow:commandBuffer];

    _GBufferRenderPassDescriptor.depthAttachment.texture = self.view.depthStencilTexture;
    _GBufferRenderPassDescriptor.stencilAttachment.texture = self.view.depthStencilTexture;

    id<MTLRenderCommandEncoder> renderEncoder = [commandBuffer renderCommandEncoderWithDescriptor:_GBufferRenderPassDescriptor];
    renderEncoder.label = @"GBuffer Generation";

    [super drawGBuffer:renderEncoder];

    [renderEncoder endEncoding];

    // Commit commands so that Metal can begin working on non-drawable dependant work without
    // waiting for a drawable to become avaliable
    [commandBuffer commit];

    commandBuffer = [self beginDrawableCommands];
    commandBuffer.label = @"Lighting Commands";

    id<MTLTexture> drawableTexture = self.currentDrawableTexture;

    // The final pass can only render if a drawable is available, otherwise it needs to skip
    // rendering this frame.
    if(drawableTexture)
    {
        // Render the lighting and composition pass

        _finalRenderPassDescriptor.colorAttachments[0].texture = drawableTexture;
        _finalRenderPassDescriptor.depthAttachment.texture = self.view.depthStencilTexture;
        _finalRenderPassDescriptor.stencilAttachment.texture = self.view.depthStencilTexture;

        id<MTLRenderCommandEncoder> renderEncoder = [commandBuffer renderCommandEncoderWithDescriptor:_finalRenderPassDescriptor];
        renderEncoder.label = @"Lighting & Composition Pass";

        [self drawDirectionalLight:renderEncoder];

        [super drawPointLightMask:renderEncoder];

        [self drawPointLights:renderEncoder];

        [super drawSky:renderEncoder];

        [super drawFairies:renderEncoder];

        [renderEncoder endEncoding];
    }

    [self endFrame:commandBuffer];
}

#if SUPPORT_BUFFER_EXAMINATION
/// Enable (or disable) buffer examination mode
- (void)validateBufferExaminationMode
{
    if(self.bufferExaminationManager.mode )
    {
        // Clear the background of the GBuffer when examining buffers.  When rendering normally
        // clearing is wasteful, but when examining the buffers, the backgrounds appear corrupt
        // making unclear what's actually rendered to the buffers
        _GBufferRenderPassDescriptor.colorAttachments[AAPLRenderTargetAlbedo].loadAction = MTLLoadActionClear;
        _GBufferRenderPassDescriptor.colorAttachments[AAPLRenderTargetNormal].loadAction = MTLLoadActionClear;
        _GBufferRenderPassDescriptor.colorAttachments[AAPLRenderTargetDepth].loadAction = MTLLoadActionClear;

        // Store depth and stencil buffers after filling them.  This is wasteful when rendering
        // normally, but necessary to present the light mask culling view.
        _finalRenderPassDescriptor.stencilAttachment.storeAction = MTLStoreActionStore;
        _finalRenderPassDescriptor.depthAttachment.storeAction = MTLStoreActionStore;
    }
    else
    {
        // When exiting buffer examination mode, return to efficient state settings
        _finalRenderPassDescriptor.stencilAttachment.storeAction = MTLStoreActionDontCare;
        _finalRenderPassDescriptor.depthAttachment.storeAction = MTLStoreActionDontCare;
        _GBufferRenderPassDescriptor.colorAttachments[AAPLRenderTargetAlbedo].loadAction = MTLLoadActionDontCare;
        _GBufferRenderPassDescriptor.colorAttachments[AAPLRenderTargetNormal].loadAction = MTLLoadActionDontCare;
        _GBufferRenderPassDescriptor.colorAttachments[AAPLRenderTargetDepth].loadAction = MTLLoadActionDontCare;
    }
}

#endif // END SUPPORT_BUFFER_EXAMINATION

@end

